jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -647,6 +647,135 @@ ecma_builtin_typedarray_prototype_reverse (ecma_value_t this_arg) /**< this argu
+                                        ecma_value_t arr_val, /**< typedarray object */
+                                        ecma_value_t offset_val) /**< offset value */
+{
+  /* 6.~ 8. targetOffset */
+  ecma_value_t ret_val = ECMA_VALUE_EMPTY;
+  ECMA_OP_TO_NUMBER_TRY_CATCH (target_offset_num, offset_val, ret_val);
+  if (ecma_number_is_nan (target_offset_num))
+  {
+    target_offset_num = 0;
+  }
+  if (target_offset_num <= -1.0 || target_offset_num >= (ecma_number_t) UINT32_MAX + 0.5)
+  {
+    return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid offset"));
+  }
+  ecma_object_t *target_typedarray_p = ecma_get_object_from_value (this_arg);
+  ecma_object_t *src_typedarray_p = ecma_get_object_from_value (arr_val);
+
+  /* 9. targetBuffer */
+  ecma_object_t *target_arraybuffer_p = ecma_typedarray_get_arraybuffer (target_typedarray_p);
+  lit_utf8_byte_t *target_buffer_p = ecma_typedarray_get_buffer (target_typedarray_p);
+
+  /* 11. targetLength */
+  ecma_length_t target_length = ecma_typedarray_get_length (target_typedarray_p);
+
+  /* 12. srcBuffer */
+  ecma_object_t *src_arraybuffer_p = ecma_typedarray_get_arraybuffer (src_typedarray_p);
+  lit_utf8_byte_t *src_buffer_p = ecma_typedarray_get_buffer (src_typedarray_p);
+
+  /* 15. targetType */
+  lit_magic_string_id_t target_class_id = ecma_object_get_class_name (target_typedarray_p);
+
+  /* 16. targetElementSize */
+  uint8_t target_shift = ecma_typedarray_get_element_size_shift (target_typedarray_p);
+  uint8_t target_element_size = (uint8_t) (1 << target_shift);
+
+  /* 17. targetByteOffset */
+  ecma_length_t target_byte_offset = ecma_typedarray_get_offset (target_typedarray_p);
+
+  /* 19. srcType */
+  lit_magic_string_id_t src_class_id = ecma_object_get_class_name (src_typedarray_p);
+
+  /* 20. srcElementSize */
+  uint8_t src_shift = ecma_typedarray_get_element_size_shift (src_typedarray_p);
+  uint8_t src_element_size = (uint8_t) (1 << src_shift);
+
+  /* 21. srcLength */
+  ecma_length_t src_length = ecma_typedarray_get_length (src_typedarray_p);
+  uint32_t src_length_uint32 = ecma_number_to_uint32 (src_length);
+  if ((ecma_number_t) src_length_uint32 != src_length)
+  {
+    return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid source length"));
+  }
+
+  /* 22. srcByteOffset */
+  ecma_length_t src_byte_offset = ecma_typedarray_get_offset (src_typedarray_p);
+
+  /* 23. */
+  uint32_t target_offset_uint32 = ecma_number_to_uint32 (target_offset_num);
+  if ((int64_t) src_length_uint32 + target_offset_uint32 > target_length)
+  {
+    ret_val = ecma_raise_range_error (ECMA_ERR_MSG ("Invalid range of index"));
+  }
+
+  /* 24.d, 25. srcByteIndex */
+  ecma_length_t src_byte_index;
+  ecma_object_t *src_arraybuffer_clone_p = NULL;
+  bool src_target_same_value = ecma_op_same_value (ecma_make_object_value (src_arraybuffer_p),
+                                                   ecma_make_object_value (target_arraybuffer_p));
+  if (src_target_same_value)
```

Comment:
That's true, I could probably just copy the buffer and operate on that instead of cloning an entire typedarray. I'll give it a try.

Reply:
I've tried to do this but I'm running into an assertion error. The code I have looks like:
```c
size_t target_clone_heap_size = (sizeof (lit_utf8_byte_t)) * target_element_size * target_length;
lit_utf8_byte_t *target_clone_buffer_p = (lit_utf8_byte_t *) jmem_heap_alloc_block (target_clone_heap_size);
memmove (target_clone_buffer_p, ecma_arraybuffer_get_buffer (target_arraybuffer_p), target_length);
src_buffer_p = target_clone_buffer_p + src_byte_offset;
// ...
```
And the assertion error is:
```c
ICE: Assertion '(uintptr_t) data_space_p % JMEM_ALIGNMENT == 0' failed at /Users/acalandra/jerryscript/jerry-core/jmem/jmem-heap.c(jmem_heap_alloc_block_internal):335.
Error: ERR_FAILED_INTERNAL_ASSERTION
```
Any idea how I can resolve this?

----
I would almost prefer the clone approach because it's closer to the spec and the code definitely looks cleaner.

Reply:
I agree that this seems like a nop. When I read the spec it added the cloning step because of side-effects. I've been trying to think of a case where this would be an issue but I haven't come up with an example.